File: packages/swt3-ai-ts/src/adapters/bedrock.ts
/**
 * SWT3 AI Witness SDK — AWS Bedrock Adapter (ES6 Proxy).
 *
 * Wraps the @aws-sdk/client-bedrock-runtime client so that
 * `client.send(new ConverseCommand(...))` and
 * `client.send(new InvokeModelCommand(...))` are intercepted.
 *
 * AWS Bedrock uses the Command pattern:
 *   const response = await client.send(new ConverseCommand({ modelId, messages }));
 *
 * The adapter intercepts `send()` and checks if the command is a
 * ConverseCommand or InvokeModelCommand to extract factors.
 *
 * Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
 */

import { sha256Truncated } from "../fingerprint.js";
import type { InferenceRecord } from "../types.js";
import type { Witness } from "../witness.js";

/**
 * Wrap a BedrockRuntimeClient with transparent witnessing.
 *
 * Usage:
 *   import { BedrockRuntimeClient, ConverseCommand } from "@aws-sdk/client-bedrock-runtime";
 *   import { Witness } from "@tenova/swt3-ai";
 *
 *   const witness = new Witness({ endpoint, apiKey, tenantId });
 *   const client = witness.wrap(new BedrockRuntimeClient({ region: "us-east-1" }));
 *
 *   const response = await client.send(new ConverseCommand({
 *     modelId: "anthropic.claude-3-5-sonnet-20241022-v2:0",
 *     messages: [{ role: "user", content: [{ text: "Hello" }] }],
 *   }));
 */
export function wrapBedrock(client: unknown, witness: Witness): unknown {
  return new Proxy(client as object, {
    get(target: object, prop: string | symbol): unknown {
      if (typeof prop === "symbol") return Reflect.get(target, prop);
      const real = Reflect.get(target, prop);

      if (prop === "send" && typeof real === "function") {
        return createSendInterceptor(real as (...args: unknown[]) => Promise<unknown>, witness, target);
      }

      return real;
    },
  });
}

function createSendInterceptor(
  realSend: (...args: unknown[]) => Promise<unknown>,
  witness: Witness,
  thisArg: object,
): (...args: unknown[]) => Promise<unknown> {
  return async function interceptedSend(...args: unknown[]): Promise<unknown> {
    const command = args[0] as Record<string, unknown> | undefined;
    if (!command) return realSend.call(thisArg, ...args);

    // Detect command type by constructor name
    const commandName = command.constructor?.name ?? "";

    if (commandName === "ConverseCommand") {
      return handleConverse(realSend, thisArg, command, witness, args);
    }

    if (commandName === "InvokeModelCommand") {
      return handleInvokeModel(realSend, thisArg, command, witness, args);
    }

    // Not a command we intercept — pass through
    return realSend.call(thisArg, ...args);
  };
}

// ── Converse Command Handler ──────────────────────────────────────

async function handleConverse(
  realSend: (...args: unknown[]) => Promise<unknown>,
  thisArg: object,
  command: Record<string, unknown>,
  witness: Witness,
  args: unknown[],
): Promise<unknown> {
  const input = (command as { input?: Record<string, unknown> }).input ?? command;
  const modelId = (input.modelId as string) ?? "unknown";
  const messages = input.messages as unknown[] ?? [];

  const promptText = extractConversePrompt(messages);
  const promptHash = sha256Truncated(promptText);

  const start = performance.now();
  const response = await realSend.call(thisArg, ...args);
  const elapsedMs = Math.round(performance.now() - start);

  const resp = response as Record<string, unknown>;
  const record = extractConverseRecord(resp, modelId, promptHash, elapsedMs);
  witness.record(record);

  return response;
}

// ── InvokeModel Command Handler ──────────────────────────────────

async function handleInvokeModel(
  realSend: (...args: unknown[]) => Promise<unknown>,
  thisArg: object,
  command: Record<string, unknown>,
  witness: Witness,
  args: unknown[],
): Promise<unknown> {
  const input = (command as { input?: Record<string, unknown> }).input ?? command;
  const modelId = (input.modelId as string) ?? "unknown";

  // Parse the body to get prompt
  let body: Record<string, unknown> = {};
  const rawBody = input.body;
  if (typeof rawBody === "string") {
    try { body = JSON.parse(rawBody); } catch { /* empty */ }
  } else if (rawBody instanceof Uint8Array) {
    try { body = JSON.parse(new TextDecoder().decode(rawBody)); } catch { /* empty */ }
  }

  const promptText = extractInvokePrompt(body, modelId);
  const promptHash = sha256Truncated(promptText);

  const start = performance.now();
  const response = await realSend.call(thisArg, ...args);
  const elapsedMs = Math.round(performance.now() - start);

  const resp = response as Record<string, unknown>;
  const record = extractInvokeRecord(resp, modelId, promptHash, elapsedMs);
  witness.record(record);

  return response;
}

// ── Prompt Extraction ─────────────────────────────────────────────

function extractConversePrompt(messages: unknown[]): string {
  const parts: string[] = [];
  for (const msg of messages) {
    const m = msg as Record<string, unknown>;
    const content = m.content;
    if (Array.isArray(content)) {
      for (const block of content) {
        const b = block as Record<string, unknown>;
        if (b.text) parts.push(b.text as string);
      }
    } else if (typeof content === "string") {
      parts.push(content);
    }
  }
  return parts.join("\n");
}

function extractInvokePrompt(body: Record<string, unknown>, modelId: string): string {
  // Anthropic on Bedrock
  const messages = body.messages as unknown[] | undefined;
  if (messages?.length) {
    const parts: string[] = [];
    for (const msg of messages) {
      const m = msg as Record<string, unknown>;
      const content = m.content;
      if (typeof content === "string") {
        parts.push(content);
      } else if (Array.isArray(content)) {
        for (const block of content) {
          const b = block as Record<string, unknown>;
          if (b.type === "text" && b.text) parts.push(b.text as string);
        }
      }
    }
    if (parts.length) return parts.join("\n");
  }

  // Amazon Titan
  if (body.inputText) return body.inputText as string;

  // Meta Llama
  if (body.prompt) return body.prompt as string;

  return JSON.stringify(body);
}

// ── Record Extraction ─────────────────────────────────────────────

function extractConverseRecord(
  response: Record<string, unknown>,
  modelId: string,
  promptHash: string,
  elapsedMs: number,
): InferenceRecord {
  let responseText = "";
  let hasRefusal = false;

  const output = response.output as Record<string, unknown> | undefined;
  const message = output?.message as Record<string, unknown> | undefined;
  const content = message?.content as Array<Record<string, unknown>> | undefined;

  if (content) {
    for (const block of content) {
      if (block.text) responseText += block.text as string;
    }
  }

  const stopReason = response.stopReason as string | undefined;
  if (stopReason === "content_filtered") hasRefusal = true;

  const usage = response.usage as Record<string, unknown> | undefined;
  const inputTokens = usage?.inputTokens as number | undefined;
  const outputTokens = usage?.outputTokens as number | undefined;

  return {
    modelId: modelId,
    modelHash: sha256Truncated(modelId),
    promptHash: promptHash,
    responseHash: sha256Truncated(responseText || ""),
    latencyMs: elapsedMs,
    inputTokens: inputTokens ?? undefined,
    outputTokens: outputTokens ?? undefined,
    hasRefusal: hasRefusal,
    provider: "bedrock",
    systemFingerprint: undefined,
    guardrailsActive: 0,
    guardrailsRequired: 0,
    guardrailPassed: true,
    guardrailNames: [],
  };
}

function extractInvokeRecord(
  response: Record<string, unknown>,
  modelId: string,
  promptHash: string,
  elapsedMs: number,
): InferenceRecord {
  let responseText = "";
  let hasRefusal = false;
  let inputTokens: number | null = null;
  let outputTokens: number | null = null;

  try {
    let body: Record<string, unknown> = {};
    const rawBody = response.body;
    if (rawBody instanceof Uint8Array) {
      body = JSON.parse(new TextDecoder().decode(rawBody));
    } else if (typeof rawBody === "string") {
      body = JSON.parse(rawBody);
    }

    // Anthropic on Bedrock
    if (body.content) {
      const content = body.content;
      if (Array.isArray(content)) {
        for (const block of content as Array<Record<string, unknown>>) {
          if (block.type === "text" && block.text) responseText += block.text as string;
        }
      } else if (typeof content === "string") {
        responseText = content;
      }
    }
    // Amazon Titan
    else if (body.results) {
      const results = body.results as Array<Record<string, unknown>>;
      if (results[0]?.outputText) responseText = results[0].outputText as string;
    }
    // Meta Llama
    else if (body.generation) {
      responseText = body.generation as string;
    }

    // Usage
    const usage = body.usage as Record<string, unknown> | undefined;
    if (usage) {
      inputTokens = (usage.input_tokens ?? usage.inputTokens ?? null) as number | null;
      outputTokens = (usage.output_tokens ?? usage.outputTokens ?? null) as number | null;
    }

    // Refusal
    const stopReason = (body.stop_reason ?? body.stopReason ?? "") as string;
    if (stopReason.toLowerCase().includes("content_filter")) hasRefusal = true;
  } catch {
    // Parse failure — still witness with what we have
  }

  return {
    modelId: modelId,
    modelHash: sha256Truncated(modelId),
    promptHash: promptHash,
    responseHash: sha256Truncated(responseText || ""),
    latencyMs: elapsedMs,
    inputTokens: inputTokens ?? undefined,
    outputTokens: outputTokens ?? undefined,
    hasRefusal: hasRefusal,
    provider: "bedrock",
    systemFingerprint: undefined,
    guardrailsActive: 0,
    guardrailsRequired: 0,
    guardrailPassed: true,
    guardrailNames: [],
  };
}

File: packages/swt3-ai/swt3_ai/adapters/bedrock.py
"""SWT3 AI Witness SDK — AWS Bedrock Adapter (ChainProxy).

Wraps the boto3 Bedrock Runtime client so that `client.converse()` and
`client.invoke_model()` are intercepted for witnessing.

AWS Bedrock API patterns:
    - converse(): High-level chat API (similar to OpenAI/Anthropic)
        client.converse(modelId="anthropic.claude-3-5-sonnet...", messages=[...])
        Response: { "output": { "message": { "content": [...] } }, "usage": {...} }

    - invoke_model(): Low-level API (raw JSON body)
        client.invoke_model(modelId="...", body=json.dumps({...}))
        Response: { "body": StreamingBody, "contentType": "application/json" }

The adapter intercepts both and extracts factors for witnessing.
The developer's code sees zero difference from using the raw client.

Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
"""

from __future__ import annotations

import json
import time
from typing import Any, Callable, Set, TYPE_CHECKING

from ..types import InferenceRecord
from ..fingerprint import sha256_truncated

if TYPE_CHECKING:
    from ..witness import Witness

# Methods to intercept on the Bedrock Runtime client
INTERCEPTED_METHODS: Set[str] = {"converse", "invoke_model"}


def wrap_bedrock(client: Any, witness: "Witness") -> "_BedrockProxy":
    """Wrap a boto3 Bedrock Runtime client with transparent witnessing.

    Usage:
        import boto3
        from swt3_ai import Witness

        witness = Witness(endpoint=..., api_key=..., tenant_id=...)
        bedrock = boto3.client("bedrock-runtime", region_name="us-east-1")
        client = witness.wrap(bedrock)

        # Use exactly as before
        response = client.converse(
            modelId="anthropic.claude-3-5-sonnet-20241022-v2:0",
            messages=[{"role": "user", "content": [{"text": "Hello"}]}],
        )
    """
    return _BedrockProxy(client, witness)


class _BedrockProxy:
    """Proxy that intercepts converse() and invoke_model() on the Bedrock client."""

    __slots__ = ("_target", "_witness")

    def __init__(self, target: Any, witness: "Witness") -> None:
        object.__setattr__(self, "_target", target)
        object.__setattr__(self, "_witness", witness)

    def __getattr__(self, name: str) -> Any:
        target = object.__getattribute__(self, "_target")
        witness = object.__getattribute__(self, "_witness")
        real_attr = getattr(target, name)

        if name in INTERCEPTED_METHODS and callable(real_attr):
            if name == "converse":
                return _make_converse_interceptor(real_attr, witness)
            elif name == "invoke_model":
                return _make_invoke_model_interceptor(real_attr, witness)

        return real_attr

    def __repr__(self) -> str:
        target = object.__getattribute__(self, "_target")
        return f"<WitnessProxy({type(target).__name__})>"


def _make_converse_interceptor(
    real_method: Callable[..., Any],
    witness: "Witness",
) -> Callable[..., Any]:
    """Intercept bedrock_runtime.converse() calls."""

    def interceptor(*args: Any, **kwargs: Any) -> Any:
        # Extract prompt from Bedrock Converse message format
        model_id = kwargs.get("modelId", "unknown")
        messages = kwargs.get("messages", [])
        prompt_text = _extract_converse_prompt(messages)
        prompt_hash = sha256_truncated(prompt_text)

        # Call and measure
        start = time.monotonic()
        response = real_method(*args, **kwargs)
        elapsed_ms = int((time.monotonic() - start) * 1000)

        # Extract record
        record = _extract_converse_record(response, model_id, prompt_hash, elapsed_ms)
        witness.record(record)

        return response

    interceptor.__name__ = "converse"
    interceptor.__doc__ = getattr(real_method, "__doc__", None)
    return interceptor


def _make_invoke_model_interceptor(
    real_method: Callable[..., Any],
    witness: "Witness",
) -> Callable[..., Any]:
    """Intercept bedrock_runtime.invoke_model() calls."""

    def interceptor(*args: Any, **kwargs: Any) -> Any:
        model_id = kwargs.get("modelId", "unknown")

        # Parse the request body to extract prompt
        body_raw = kwargs.get("body", "{}")
        if isinstance(body_raw, (bytes, bytearray)):
            body_raw = body_raw.decode("utf-8")
        try:
            body = json.loads(body_raw) if isinstance(body_raw, str) else body_raw
        except (json.JSONDecodeError, TypeError):
            body = {}

        prompt_text = _extract_invoke_prompt(body, model_id)
        prompt_hash = sha256_truncated(prompt_text)

        # Call and measure
        start = time.monotonic()
        response = real_method(*args, **kwargs)
        elapsed_ms = int((time.monotonic() - start) * 1000)

        # Extract record from invoke_model response
        record = _extract_invoke_record(response, model_id, prompt_hash, elapsed_ms)
        witness.record(record)

        return response

    interceptor.__name__ = "invoke_model"
    interceptor.__doc__ = getattr(real_method, "__doc__", None)
    return interceptor


# ── Prompt Extraction ────────────────────────────────────────────────

def _extract_converse_prompt(messages: Any) -> str:
    """Extract hashable text from Bedrock Converse message format.

    Bedrock Converse format:
        [{"role": "user", "content": [{"text": "Hello"}]}]
    """
    parts: list[str] = []
    if isinstance(messages, (list, tuple)):
        for msg in messages:
            if isinstance(msg, dict):
                content = msg.get("content", [])
                if isinstance(content, list):
                    for block in content:
                        if isinstance(block, dict):
                            text = block.get("text", "")
                            if text:
                                parts.append(text)
                elif isinstance(content, str):
                    parts.append(content)
    return "\n".join(parts)


def _extract_invoke_prompt(body: dict, model_id: str) -> str:
    """Extract prompt from invoke_model body (varies by model provider).

    Anthropic on Bedrock:
        {"messages": [{"role": "user", "content": "Hello"}], ...}

    Amazon Titan:
        {"inputText": "Hello", ...}

    Meta Llama:
        {"prompt": "Hello", ...}
    """
    # Anthropic format (most common on Bedrock)
    messages = body.get("messages", [])
    if messages:
        parts: list[str] = []
        for msg in messages:
            if isinstance(msg, dict):
                content = msg.get("content", "")
                if isinstance(content, str):
                    parts.append(content)
                elif isinstance(content, list):
                    for block in content:
                        if isinstance(block, dict) and block.get("type") == "text":
                            parts.append(block.get("text", ""))
        if parts:
            return "\n".join(parts)

    # Amazon Titan format
    if "inputText" in body:
        return str(body["inputText"])

    # Meta Llama format
    if "prompt" in body:
        return str(body["prompt"])

    return json.dumps(body)


# ── Record Extraction ────────────────────────────────────────────────

def _extract_converse_record(
    response: Any,
    model_id: str,
    prompt_hash: str,
    elapsed_ms: int,
) -> InferenceRecord:
    """Extract InferenceRecord from Bedrock Converse response.

    Response shape:
        {
            "output": {"message": {"role": "assistant", "content": [{"text": "..."}]}},
            "stopReason": "end_turn",
            "usage": {"inputTokens": 10, "outputTokens": 20},
            "metrics": {"latencyMs": 500}
        }
    """
    response_text = ""
    has_refusal = False

    output = response.get("output", {})
    message = output.get("message", {})
    content = message.get("content", [])

    if isinstance(content, list):
        for block in content:
            if isinstance(block, dict):
                text = block.get("text", "")
                if text:
                    response_text += text

    stop_reason = response.get("stopReason", "")
    if stop_reason == "content_filtered":
        has_refusal = True

    # Token usage
    usage = response.get("usage", {})
    input_tokens = usage.get("inputTokens")
    output_tokens = usage.get("outputTokens")

    # Model hash
    model_hash = sha256_truncated(model_id)
    response_hash = sha256_truncated(response_text) if response_text else sha256_truncated("")

    return InferenceRecord(
        model_id=model_id,
        model_hash=model_hash,
        prompt_hash=prompt_hash,
        response_hash=response_hash,
        latency_ms=elapsed_ms,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        has_refusal=has_refusal,
        provider="bedrock",
        system_fingerprint=None,
    )


def _extract_invoke_record(
    response: Any,
    model_id: str,
    prompt_hash: str,
    elapsed_ms: int,
) -> InferenceRecord:
    """Extract InferenceRecord from Bedrock invoke_model response.

    Response has a StreamingBody in response["body"] that must be read.
    """
    response_text = ""
    has_refusal = False
    input_tokens = None
    output_tokens = None

    try:
        body_bytes = response.get("body", b"")
        if hasattr(body_bytes, "read"):
            body_bytes = body_bytes.read()
        body = json.loads(body_bytes) if body_bytes else {}

        # Anthropic on Bedrock response
        if "content" in body:
            content = body["content"]
            if isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and block.get("type") == "text":
                        response_text += block.get("text", "")
            elif isinstance(content, str):
                response_text = content

        # Amazon Titan response
        elif "results" in body:
            results = body["results"]
            if isinstance(results, list) and results:
                response_text = results[0].get("outputText", "")

        # Meta Llama response
        elif "generation" in body:
            response_text = body.get("generation", "")

        # Token usage (Anthropic format)
        usage = body.get("usage", {})
        input_tokens = usage.get("input_tokens") or usage.get("inputTokens")
        output_tokens = usage.get("output_tokens") or usage.get("outputTokens")

        # Refusal check
        stop_reason = body.get("stop_reason", "") or body.get("stopReason", "")
        if "content_filter" in str(stop_reason).lower():
            has_refusal = True

    except (json.JSONDecodeError, AttributeError, TypeError):
        pass

    model_hash = sha256_truncated(model_id)
    response_hash = sha256_truncated(response_text) if response_text else sha256_truncated("")

    return InferenceRecord(
        model_id=model_id,
        model_hash=model_hash,
        prompt_hash=prompt_hash,
        response_hash=response_hash,
        latency_ms=elapsed_ms,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        has_refusal=has_refusal,
        provider="bedrock",
        system_fingerprint=None,
    )
